src/irmin-pack/inode.ml

```diff
@@ -259,27 +255,116 @@ struct
       |> sealr
   end
 
+  (** [Val_impl] defines the structure logic of inodes.
+
+      {3 Inode layout}
+
+      The layout ['a layout] associated to an inode ['a t] defines certain
+      properties of the data structure:
+
+      - When [Total], the inode is self contained and immutable.
+      - When [Partial], chunks of the inode might be missing but they can be
+        fetched from the backend when needed. Mutable pointers act as cache.
+      - When [Truncated], chunks of the inode might be missing. Those chunks are
+        unreachable because the pointer to the backend is missing. The inode is
+        immutable.
+
+      The layout is of an inode is determined from the module [Val], it depends
+      on the way the inode was constructed:
+
+      - When [Total], it originates from [Val.v] or [Val.empty].
+      - When [Partial], it originates from [Val.of_bin], which is only used by
+        [Inode.find].
+      - When [Truncated], it originates from an [Irmin.Type] unserialisation
+        made possible by [Val.t].
+
+      Most functions in [Val_impl] are polymorphic regarding the layout of the
+      manipulated inode. *)
   module Val_impl = struct
     open T
 
-    let equal_hash = Irmin.Type.(unstage (equal hash_t))
     let equal_value = Irmin.Type.(unstage (equal value_t))
 
-    type ptr = { target_hash : hash Lazy.t; mutable target : t option }
-
-    and tree = { depth : int; length : int; entries : ptr option array }
+    type _ layout =
+      | Total : direct_ptr layout
+      | Partial : (hash -> lazy_ptr t option) -> lazy_ptr layout
+      | Truncated : flaky_ptr layout
 
-    and v = Values of value StepMap.t | Tree of tree
-
-    and t = { hash : hash Lazy.t; stable : bool; v : v }
+    and lazy_ptr = {
+      target_hash : hash Lazy.t;
+      mutable target : lazy_ptr t option;
+    }
 
-    let pred t =
+    and direct_ptr = Direct of direct_ptr t [@@unboxed]
+
+    and flaky_ptr = Broken of hash | Intact of flaky_ptr t
+
+    and 'ptr tree = { depth : int; length : int; entries : 'ptr option array }
+
+    and 'ptr v = Values of value StepMap.t | Tree of 'ptr tree
+
+    and 'ptr t = { hash : hash Lazy.t; stable : bool; v : 'ptr v }
+
+    module Ptr = struct
+      let hash : type ptr. ptr layout -> ptr -> _ = function
+        | Total -> fun (Direct ptr) -> Lazy.force ptr.hash
+        | Partial _ -> fun { target_hash; _ } -> Lazy.force target_hash
+        | Truncated -> (
+            function Broken h -> h | Intact ptr -> Lazy.force ptr.hash)
+
+      let target : type ptr. ptr layout -> ptr -> ptr t =
+       fun la ->
```

Comment:
maybe calling this variable `layout` is more suggestive

Reply:
I wanted something short, but something long is okey too